### PR TITLE
Fix Colab logger error

### PR DESCRIPTION
**Pull request by qgallouedec**
## Description

Related : https://github.com/python/mypy/issues/15191

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

closes https://github.com/DLR-RM/stable-baselines3/issues/1483

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [ ] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [ ] I have reformatted the code using `make format` (**required**)
- [ ] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)
- [ ] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: docs/misc/changelog.rst [hunk 1]
```diff
@@ -37,6 +37,7 @@ Bug Fixes:
 - Fixed ``VecExtractDictObs`` does not handle terminal observation (@WeberSamuel)
 - Set NumPy version to ``>=1.20`` due to use of ``numpy.typing`` (@troiganto)
 - Fixed loading DQN changes ``target_update_interval`` (@tobirohrer)
+- Fixed logger for Google Colab
```
Comment (qgallouedec) [inline docs/misc/changelog.rst:40]: @araffin is it worth mentioning since the bug has been introduced by 2.0.0a7?

Comment (araffin) [reply docs/misc/changelog.rst:40]: no, but I would rather mention a breaking change in the way the logger check the file

File: stable_baselines3/common/logger.py [hunk 2]
```diff
@@ -159,12 +159,12 @@ class HumanOutputFormat(KVWriter, SeqWriter):
         no longer than 79 characters wide.
     """
 
-    def __init__(self, filename_or_file: Union[str, TextIO], max_length: int = 36):
+    def __init__(self, filename_or_file: Union[str, TextIOBase], max_length: int = 36):
```
Comment (araffin) [inline stable_baselines3/common/logger.py:162]: Looking at https://stackoverflow.com/questions/61000227/type-checking-issue-with-io-textiobase-in-a-union and https://github.com/python/typeshed/pull/10135/files, why not `TextIO` for the annotation
and `TextIOBase` for the check?

```python
import io
import sys
import typing


def test(out: typing.TextIO) -> None:
    assert isinstance(out, io.TextIOBase)
    out.write("test")


test(sys.stdout)

file = open("test.text", "w")

test(file)
```

File: docs/misc/changelog.rst (hunk 1 above)
Comment (araffin) [reply docs/misc/changelog.rst:40]: It would also be nice to add a regression test if possible

File: stable_baselines3/common/logger.py (hunk 2 above)
Comment (qgallouedec) [reply stable_baselines3/common/logger.py]: Let's try this (so unintuitive...)

File: docs/misc/changelog.rst (hunk 1 above)
Comment (qgallouedec) [reply docs/misc/changelog.rst:40]: How would you do that? By using the `OutStream` of `ipykernel` for a test?

**Comment by qgallouedec** [comment]
Now it works on colab

File: docs/misc/changelog.rst (hunk 1 above)
Comment (araffin) [reply docs/misc/changelog.rst]: not sure, in that case I would create a dummy subclass of text io base and override stdout (that's what ipykernel does if get it)?
(best would be with ipykernel indeed, but I don't think it's part of SB3 dependencies)

Comment (qgallouedec) [reply docs/misc/changelog.rst]: > I don't think it's part of SB3 dependencies

Agree

> create a dummy subclass of text io base

check https://github.com/DLR-RM/stable-baselines3/pull/1484/commits/45db2c14b71b7cfc09ae21b69805520d8443ac50

File: tests/test_logger.py [hunk 3]
```diff
@@ -434,3 +435,44 @@ def test_ep_buffers_stats_window_size(algo, stats_window_size):
...
+        def close(self) -> None:
+            pass
+
+        def get_printed(self) -> str:
+            return "\n".join(["".join(line) for line in self.lines])
+
+    dummy_text_io = DummyTextIO()
+    o = HumanOutputFormat(dummy_text_io)
+    o.write({"key1": "value1", "key2": 42}, {"key1": None, "key2": None})
+    o.write({"key1": "value2", "key2": 43}, {"key1": None, "key2": None})
```
Comment (araffin) [inline tests/test_logger.py:461]:
```suggestion
    dummy_text_io = DummyTextIO()
    output = HumanOutputFormat(dummy_text_io)
    output.write({"key1": "value1", "key2": 42}, {"key1": None, "key2": None})
    output.write({"key1": "value2", "key2": 43}, {"key1": None, "key2": None})
```

File: tests/test_logger.py [hunk 4]
```diff
@@ -434,3 +435,44 @@ def test_ep_buffers_stats_window_size(algo, stats_window_size):
...
+| key1 | value2 |
+| key2 | 43     |
+-----------------
+
+"""
+
+    assert printed == desired_printed
+
+
+test_human_output_format_custom_test_io()
```
Comment (araffin) [inline tests/test_logger.py]:
```suggestion
```
you can use `pytest -v tests/ -k human_output_format_custom_test_io` if you want to try a single test

Comment (qgallouedec) [reply tests/test_logger.py]: It's to use the debugger 😉

**Comment by qgallouedec** [comment]
The new test indeed fails with 2.0.0a7.

**Review by araffin — APPROVED** [review]
LGTM, thanks =)

**Pull request merged**